detectors/resources-support/src/main/java/com/google/cloud/opentelemetry/detectors/GoogleAppEngine.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.opentelemetry.detectors;
+
+import static com.google.cloud.opentelemetry.detectors.AttributeKeys.GAE_APP_VERSION;
+import static com.google.cloud.opentelemetry.detectors.AttributeKeys.GAE_AVAILABILITY_ZONE;
+import static com.google.cloud.opentelemetry.detectors.AttributeKeys.GAE_CLOUD_REGION;
+import static com.google.cloud.opentelemetry.detectors.AttributeKeys.GAE_INSTANCE_ID;
+import static com.google.cloud.opentelemetry.detectors.AttributeKeys.GAE_MODULE_NAME;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+final class GoogleAppEngine implements DetectedPlatform {
+  private final EnvironmentVariables environmentVariables;
+  private final GCPMetadataConfig metadataConfig;
+  private final Map<String, Optional<String>> availableAttributes;
+
+  GoogleAppEngine() {
+    this.environmentVariables = EnvironmentVariables.DEFAULT_INSTANCE;
+    this.metadataConfig = GCPMetadataConfig.DEFAULT_INSTANCE;
+    this.availableAttributes = prepareAttributes();
+  }
+
+  // for testing only
+  GoogleAppEngine(EnvironmentVariables environmentVariables, GCPMetadataConfig metadataConfig) {
+    this.environmentVariables = environmentVariables;
+    this.metadataConfig = metadataConfig;
+    this.availableAttributes = prepareAttributes();
+  }
+
+  private Map<String, Optional<String>> prepareAttributes() {
```

Comment:
Returning `Map<String, Optional<String>>` is fairly uncommon in Java.  In functional (newer) languages, like Kotlin or Scala, you'd expect `Map` to support a `get` that returns an `Optional` value, denoting non-existence.

Here, I'd suggest still using `Map<String,String>` and on the read side doing:


```java
Optional.ofNullable(map.get(key))
```

This will still let you use your `ifPresent` trick and is less prone to runtime exceptions than what you're doing now (in the event your keys do not exist in the map).

Reply:
Sounds good 👍🏻 

I agree - the main reason behind using `Optional` was the `ifPresent` trick.